### PR TITLE
Volumetric locking tensor

### LINKED ISSUE #8235: Volumetric locking correction for Tensor mechanics
### Description of the enhancement or error report
B-bar correction has been implemented for the tensor mechanics module. This is the similar to the changes made to the solid mechanics module in #7260. In tensor mechanics module, volumetric locking correction is set to false by default. It can be turned on by specifying "volumetric_locking_correction=true" in GlobalParams. 

### Rationale for the enhancement or information for reproducing the error
To prevent locking of linear elements (Hex8 or quad4) when the material nears the incompressible limit.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
A few tests in the tensor mechanics module have been re-golded..

### PR DESCRIPTION
**Pull request by sveerara**
@bwspenc , @sapitts : There are three commit in this merge request. 

First commit has the tensor mechanics kernels and the materials code changes. Volumetric locking correction is set to false by default. In 1-D cases, an error would be shown if volumetric locking correction is set to true. 

In the second commit, new tests have been added. These tests check the volumetric locking correction implementation for the basic mechanics tests. Also one element tests have been added to check the hand coded jacobian against the finite difference jacobian. 

In the third commit, earlier tests which used finite strain calculation have been edited. The earlier implementation of finite strain calculation had volumetric locking correction turned on by default. 

closes #8235

### CODE REVIEW

**Comment by sveerara** [comment]
Bison patch is ready. 

One tests in Grizzly is failing. I don't have access to Grizzly. @bwspenc can you please take a look at the grizzly test. Setting volumetric_locking_correction = true in GlobalParams got rid of most of the Bison errors. 

Mastodon is failing because libmesh must be compiled with VTK. I am not sure why Mammoth is failing.

File: framework/include/utils/RankFourTensor.h [hunk 1]
```diff
@@ -205,6 +205,13 @@ class RankFourTensor
   /// Inner product of the major transposed tensor with a rank two tensor
   RankTwoTensor innerProductTranspose(const RankTwoTensor &) const;
 
+  /// Calculates the sum of Ciijj for i and j varying from 0 to 2
+  Real sum_3x3() const;
```
Comment (dschwen) [inline framework/include/utils/RankFourTensor.h]: This is at odds with our naming convention. I suggest removing the `_`

File: modules/tensor_mechanics/include/kernels/StressDivergenceTensors.h [hunk 2]
```diff
@@ -62,6 +65,11 @@ class StressDivergenceTensors : public ALEKernel
   const bool _temp_coupled;
 
   const unsigned int _temp_var;
+
+  /// Parameters for volumetric locking correction
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/StressDivergenceTensors.h]: please use `@{` `@}` to have the doxygen comment apply to each parameter or give each parameter its own comment. This way only the first parameter will have the given documentation and the remaining two will be undocumented.

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 3]
```diff
@@ -153,6 +153,14 @@ TensorMechanicsAction::TensorMechanicsAction(const InputParameters & params)
     std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
     _generate_output.push_back(lower);
   }
+
+  // Error if volumetric locking correction is true for 1D problems
+  if (_ndisp == 1)
+  {
+    bool volumetric_locking_correction = getParam<bool>("volumetric_locking_correction");
+    if (volumetric_locking_correction)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: do 

```
if (_ndisp == 1 && getParam<bool>("volumetric_locking_correction"))
  mooseError("Volumetric locking correction should be set to false for 1D problems.");
```

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 4]
```diff
@@ -264,6 +272,7 @@ TensorMechanicsAction::act()
     params.set<std::vector<VariableName>>("displacements") = _coupled_displacements;
     params.set<bool>("use_displaced_mesh") = false;
     params.set<std::vector<MaterialPropertyName>>("eigenstrain_names") = _eigenstrain_names;
+    params.set<bool>("volumetric_locking_correction") = getParam<bool>("volumetric_locking_correction");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: No! This line is not needed. I'm using `applyParameters`

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 5]
```diff
@@ -450,5 +459,7 @@ TensorMechanicsAction::getKernelParameters(std::string type)
     mooseDeprecated("Use 'temperature' instead of 'temp'");
   }
 
+  params.set<bool>("volumetric_locking_correction") = getParam<bool>("volumetric_locking_correction");
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: No, remove this as well (see above)

File: modules/tensor_mechanics/src/kernels/DynamicStressDivergenceTensors.C [hunk 6]
```diff
@@ -43,26 +43,40 @@ DynamicStressDivergenceTensors::computeQpResidual()
...
+  Real residual = 0.0;
   if (_static_initialization && _t == _dt)
+  {
     // If static inialization is true, then in the first step residual is only Ku which is stress.grad(test).
-    return _stress[_qp].row(_component) * _grad_test[_i][_qp];
+    residual +=  _stress[_qp].row(_component) * _grad_test[_i][_qp];
+
+    if (_volumetric_locking_correction)
+      residual += _stress[_qp].trace() / 3.0 * (_avg_grad_test[_i][_component] - _grad_test[_i][_qp](_component));
+
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/DynamicStressDivergenceTensors.C]: rm empty line

File: modules/tensor_mechanics/src/kernels/DynamicStressDivergenceTensors.C [hunk 7]
```diff
@@ -43,26 +43,40 @@ DynamicStressDivergenceTensors::computeQpResidual()
...
-    return 0;
+  {
+    residual +=  _stress[_qp].row(_component) * _grad_test[_i][_qp] * (1.0 + _alpha + (1.0 + _alpha) * _zeta / _dt)
+               - (_alpha + (1.0 + 2.0 * _alpha) * _zeta / _dt) * _stress_old[_qp].row(_component) * _grad_test[_i][_qp]
+               + (_alpha * _zeta/_dt) * _stress_older[_qp]. row(_component) * _grad_test[_i][_qp];
+
+    if (_volumetric_locking_correction)
+      residual +=  (_stress[_qp].trace() * (1.0 + _alpha + (1.0 + _alpha) * _zeta / _dt)
+               - (_alpha + (1.0 + 2.0 * _alpha) * _zeta / _dt) * _stress_old[_qp].trace()
+               + (_alpha * _zeta/_dt) * _stress_older[_qp]. trace()) / 3.0 * (_avg_grad_test[_i][_component] - _grad_test[_i][_qp](_component));
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/DynamicStressDivergenceTensors.C]: make indentation consistent

File: modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C [hunk 8]
```diff
@@ -34,19 +35,32 @@ StressDivergenceRZTensors::initialSetup()
...
-           + _grad_test[_i][_qp](1) * _stress[_qp](0,1); // stress_{rz}
+    div = _grad_test[_i][_qp](0) * _stress[_qp](0,0) +
+        +  (_test[_i][_qp] / _q_point[_qp](0)) * _stress[_qp](2,2) +
+        + _grad_test[_i][_qp](1) * _stress[_qp](0,1); // stress_{rz}
+
+    // volumetric locking correction
+    if (_volumetric_locking_correction)
+      div += (_avg_grad_test[_i][0] - _grad_test[_i][_qp](0) - _test[_i][_qp] / _q_point[_qp](0)) * (_stress[_qp].trace()) / 3.0;
+
+
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C]: rm empty lines

File: modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C [hunk 9]
```diff
@@ -106,25 +121,108 @@ StressDivergenceRZTensors::calculateJacobian(unsigned int ivar, unsigned int jva
...
+  return val / 3.0 + first_term;
+}
+
+void
+StressDivergenceRZTensors::computeAverageGradientTest()
+{
+  // calculate volume averaged value of shape function derivative
+  _avg_grad_test.resize(_test.size());
+  for (_i = 0; _i < _test.size(); _i++)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C]: we use pre increment `++_i` pretty consistently in TM

File: modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C [hunk 10]
```diff
@@ -106,25 +121,108 @@ StressDivergenceRZTensors::calculateJacobian(unsigned int ivar, unsigned int jva
...
+  {
+    _avg_grad_test[_i].resize(2);
+    _avg_grad_test[_i][_component] = 0.0;
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      if (_component == 0)
+        _avg_grad_test[_i][_component] += (_grad_test[_i][_qp](_component) + _test[_i][_qp] / _q_point[_qp](0)) * _JxW[_qp] * _coord[_qp];
+      else
+        _avg_grad_test[_i][_component] += _grad_test[_i][_qp](_component) * _JxW[_qp] * _coord[_qp];
+
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRZTensors.C]: rm empty line

**Review by dschwen — CHANGES_REQUESTED** [review]

**Comment by sveerara** [comment]
I have addressed the comments. Thanks!

**Comment by bwspenc** [comment]
So is this ready to go?  @sveerara, you said that you have a fix ready for BISON, right?  I can fix the Grizzly test diff.

**Comment by sveerara** [comment]
Yes, I have the bison fix ready.

**Review by dschwen — APPROVED** [review]

File: framework/include/utils/RankFourTensor.h [hunk 11]
```diff
@@ -205,6 +205,13 @@ class RankFourTensor
   /// Inner product of the major transposed tensor with a rank two tensor
   RankTwoTensor innerProductTranspose(const RankTwoTensor &) const;
 
+  /// Calculates the sum of Ciijj for i and j varying from 0 to 2
+  Real sum3x3() const;
+
+  /// Calculates the vector a[i] = sum over j Ciijj for i and j varying from 0 to 2
+  RealGradient sum3x1() const;
+
+
```
Comment (dschwen) [inline framework/include/utils/RankFourTensor.h]: rm extra line

File: modules/tensor_mechanics/src/kernels/StressDivergenceTensors.C [hunk 12]
```diff
@@ -131,8 +226,51 @@ StressDivergenceTensors::computeQpOffDiagJacobian(unsigned int jvar)
...
+      {
+        phi(1,1) = _grad_phi[_j][_qp](1);
+        phi(1,0) = _grad_phi[_j][_qp](0);
+        phi(1,2) = _grad_phi[_j][_qp](2);
+      }
+      else if (coupled_component == 2)
+      {
+        phi(2,2) = _grad_phi[_j][_qp](2);
+        phi(2,0) = _grad_phi[_j][_qp](0);
+        phi(2,1) = _grad_phi[_j][_qp](1);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/StressDivergenceTensors.C]: could be simplified to 
```
for (unsigned int i = 0; i < 3; ++i)
  phi(_component, i) = _grad_phi[_j][_qp](i);
```

removing 16 lines of code

**Comment by dschwen** [comment]
@bwspenc is standing by to fix Bison and Grizzly. Merging now.

**Comment by brianmoose** [comment]
This seems to have caused the heavy test `combined:j2_plasticity_vs_LSH.j2_hard1_mod` to fail with a CSVDIFF. See https://www.moosebuild.org/job/58993/

**Comment by sveerara** [comment]
@brianmoose, thanks for the information. I will correct the test and submit a merge request.

**Pull request merged**